packages/stokhos/src/sacado/kokkos/vector/Teuchos_BLAS_MP_Vector.hpp

```diff
@@ -45,99 +45,465 @@
-          r = scale * STS::squareroot (da_scaled*da_scaled + db_scaled*db_scaled);
-          r = SIGN (STS::one(), roe) * r;
-          *c = *da / r;
-          *s = *db / r;
-          z = STS::one();
-          if (STS::magnitude (*da) > STS::magnitude (*db)) {
-            z = *s;
-          }
-          if (STS::magnitude (*db) >= STS::magnitude (*da) && *c != STS::zero()) {
-            z = STS::one() / *c;
-          }
+namespace Teuchos
+{
+
+namespace details
+{
+
+template <typename Storage>
+class GivensRotator<Sacado::MP::Vector<Storage>, false>
+{
+public:
+  typedef Sacado::MP::Vector<Storage> ScalarType;
+  typedef ScalarType c_type;
+
+  void
+  ROTG(ScalarType *da,
+       ScalarType *db,
+       ScalarType *c,
+       ScalarType *s) const
+  {
+    typedef ScalarTraits<ScalarType> STS;
+
+    ScalarType r, roe, scale, z, da_scaled, db_scaled;
+    auto m_da = (STS::magnitude(*da) > STS::magnitude(*db));
+    mask_assign(m_da, roe) = {*da, *db};
+
+    scale = STS::magnitude(*da) + STS::magnitude(*db);
+
+    auto m_scale = scale != STS::zero();
+
+    da_scaled = *da;
+    db_scaled = *db;
+
+    *c = *da;
+    *s = *db;
+
+    ScalarType tmp = STS::one();
+    mask_assign(m_scale, tmp) /= scale;
+
+    mask_assign(m_scale, da_scaled) *= tmp;
+    mask_assign(m_scale, db_scaled) *= tmp;
+
+    r = scale * STS::squareroot(da_scaled * da_scaled + db_scaled * db_scaled);
+    auto m_roe = roe < 0;
+    mask_assign(m_roe, r) = -r;
+
+    tmp = STS::one();
+    mask_assign(m_scale, tmp) /= r;
+
+    mask_assign(m_scale, *c) *= tmp;
+    mask_assign(m_scale, *s) *= tmp;
+
+    mask_assign(!m_scale, *c) = STS::one();
+    mask_assign(!m_scale, *s) = STS::zero();
+
+    mask_assign(*c != STS::zero(), z) /= {STS::one(), *c, STS::zero()};
+    mask_assign(!m_scale, z) = STS::zero();
+    mask_assign(m_da, z) = *s;
+
+    *da = r;
+    *db = z;
+  }
+};
+} // namespace details
+} // namespace Teuchos
+
+//namespace Sacado {
+//  namespace MP {
+namespace Teuchos
+{
+//! Vector specializations for Teuchos::BLAS wrappers
+template <typename OrdinalType, typename Storage>
+class BLAS<OrdinalType, Sacado::MP::Vector<Storage>> : public Teuchos::DefaultBLASImpl<OrdinalType, Sacado::MP::Vector<Storage>>
+{
+
+  typedef typename Teuchos::ScalarTraits<Sacado::MP::Vector<Storage>>::magnitudeType MagnitudeType;
+  typedef typename Sacado::ValueType<Sacado::MP::Vector<Storage>>::type ValueType;
+  typedef typename Sacado::ScalarType<Sacado::MP::Vector<Storage>>::type scalar_type;
+  typedef typename Sacado::MP::Vector<Storage> ScalarType;
+  typedef Teuchos::DefaultBLASImpl<OrdinalType, Sacado::MP::Vector<Storage>> BLASType;
+
+public:
+  //! @name Constructor/Destructor.
+  //@{
+
+  //! Default constructor.
+  BLAS(bool use_default_impl = true,
+       bool use_dynamic = true, OrdinalType static_workspace_size = 0){};
+
+  //! Copy constructor.
+
+  BLAS(const BLAS &x){};
+
+  //! Destructor.
+  virtual ~BLAS(){};
```

Comment:
1. Does BLAS have any virtual methods?  If not, then you don't need this.
2. If you do need this, do this: `virtual BLAS() = default;`